internal/provider/serverless_cluster_resource.go

```diff
@@ -0,0 +1,730 @@
+}
+
+type spendingLimit struct {
+	Monthly types.Int64 `tfsdk:"monthly"`
+}
+
+type automatedBackupPolicy struct {
+	StartTime     types.String `tfsdk:"start_time"`
+	RetentionDays types.Int64  `tfsdk:"retention_days"`
+}
+
+type endpoints struct {
+	PublicEndpoint  publicEndpoint  `tfsdk:"public_endpoint"`
+	PrivateEndpoint privateEndpoint `tfsdk:"private_endpoint"`
+}
+
+type publicEndpoint struct {
+	Host     types.String `tfsdk:"host"`
+	Port     types.Int64  `tfsdk:"port"`
+	Disabled types.Bool   `tfsdk:"disabled"`
+}
+
+type privateEndpoint struct {
+	Host        types.String `tfsdk:"host"`
+	Port        types.Int64  `tfsdk:"port"`
+	AWSEndpoint *awsEndpoint `tfsdk:"aws_endpoint"`
+	GCPEndpoint *gcpEndpoint `tfsdk:"gcp_endpoint"`
+}
+
+type awsEndpoint struct {
+	ServiceName      types.String `tfsdk:"service_name"`
+	AvailabilityZone types.List   `tfsdk:"availability_zone"`
+}
+
+type gcpEndpoint struct {
+	ServiceAttachmentName types.String `tfsdk:"service_attachment_name"`
+}
+
+type encryptionConfig struct {
+	EnhancedEncryptionEnabled types.Bool `tfsdk:"enhanced_encryption_enabled"`
+}
+
+type usage struct {
+	RequestUnit     types.String `tfsdk:"request_unit"`
+	RowBasedStorage types.Int64  `tfsdk:"row_based_storage"`
+	ColumnarStorage types.Int64  `tfsdk:"columnar_storage"`
+}
+
+type serverlessClusterResource struct {
+	provider *tidbcloudProvider
+}
+
+func NewServerlessClusterResource() resource.Resource {
+	return &serverlessClusterResource{}
+}
+
+func (r *serverlessClusterResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_serverless_cluster"
+}
+
+func (r *serverlessClusterResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	var ok bool
+	if r.provider, ok = req.ProviderData.(*tidbcloudProvider); !ok {
+		resp.Diagnostics.AddError("Internal provider error",
+			fmt.Sprintf("Error in Configure: expected %T but got %T", tidbcloudProvider{}, req.ProviderData))
+	}
+}
+
+func (r *serverlessClusterResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "serverless cluster resource",
+		Attributes: map[string]schema.Attribute{
+			"cluster_id": schema.StringAttribute{
+				MarkdownDescription: "The ID of the cluster.",
+				Computed:            true,
+			},
+			"display_name": schema.StringAttribute{
+				MarkdownDescription: "The display name of the cluster.",
+				Required:            true,
+			},
+			"region": schema.SingleNestedAttribute{
+				MarkdownDescription: "The region of the cluster.",
+				Required:            true,
+				Attributes: map[string]schema.Attribute{
+					"name": schema.StringAttribute{
+						MarkdownDescription: "The unique name of the region.",
```

Comment:
better to tell user how to get the region name

Reply:
how about let users input region_id then we concat the name